README.md
# ISO 8583 codec

## Scope

Implement encoding and decoding of ISO 8583 messages from/to buffers, usually
 to be exchanged over a TCP socket. Wikipedia has some
 [documentation](http://en.wikipedia.org/wiki/ISO_8583) for the standard.

## Usage

Usually used together with ut-port-tcp, for example like this:

```javascript
module.exports = {
    createPort: require('ut-port-tcp'),
    format: {
        size: '16/integer',
        codec: require('ut-codec-iso8583'),
        version: 0,
        baseEncoding: 'ascii',
        fieldFormat: {
            '37': {format: 'numeric'}
        }
    }
}
```

In the _format_ configuration property, the following settings are often used:

- _size_ - specifies the format for the message size header and is mandatory
  to be specified
- _version_ - specifies the ISO 8583 version to be used
- _baseEncoding_ - specifies one of the 2 most common base encodings to use:
 _ascii_ or _binary_, with _ascii_ being default
- _fieldFormat_ - in this object, overrides for the format of specific fields
  can be specified. If not specified, a default format for the _version_ and
   _baseEncoding_ will be used. The default formats are in the file
   [fields.json](./fields.json)

## Public API

The module exposes _Iso8583_ class.

### decode

(buffer, $meta, context, log)

Decodes ISO8583 format message to object

params

- _buffer_ (buffer) - ISO8583 encoded message
- _$meta_ (object) - meta object as defined in ???
- _context_ (object) - context object as defined in ???
- _log_ (object) - log functions as defined in _ut-log_

result (object)

- _header_ (string) - message header
- _mtid_ (string) - message type indicator according to ISO8583 specifications
- rawData (buffer) - message in ISO8583 format
- _0_ (string) - primary bitmap
- _1_ (string) - secondary bitmap
- __rest__ - depending on data elements _0_ and _1_ contains ISO8583 data
 elements as _key: value_ pairs as follows:
  - _key_ - ISO8583 element index
  - _value_ (string) - ISO8583 element value

### encode

(message, $meta, context, log)

Encodes message object to ISO8583 format

params

- _message_ (object) - each _key: value_ pair contains the following:
  - _key_ - ISO8583 element index
  - _value_ (string) - ISO8583 element value
- _$meta_ (object) - meta object as defined in ???
- _context_ (object) - context object as defined in ???
- _log_ (object) - log functions as defined in _ut-log_

result (buffer) - ISO8583 encoded _message_

## Field format definition

Each ISO8583 field is defined in the following object:

- _key_ - ISO8583 element index, including _header_, _footer_, _mtid_ and
 primary bitmap _0_
- _value_ (object) - data element definition
  - _size_ (integer) - size of the data element value in number of symbols
  - _name_ (string) - descriptive field name
  - _format_ (string) - data format of the value
  - _prefixSize_ (integer) - number of symbols indicating the _size_ of the
   variable length data elements; **NOTE: available only for (and also
   indicating) variable length data elements**

## Configuration

- _defineError_ (function) - defineError function as defined in _ut-error_
- _getError_ (function) - getError function as defined in _ut-error_
- _fetchErrors_ (function) - fetchErrors function as defined in _ut-error_
- _maskFields_ (array) - contains ISO8583 element indexes which will be masked
 in ISO8583 encoded messages
  - (string) - ISO8583 data element
- _networkCodes_ () - ???
- _emvParser_ () - ???
- _successResponseIdentifier_ (string) - error code for successful processing
 (no-error code)
- _version_ (string) - ISO8583 version; __Valid values__: _0_ (1987), _1_
 (1993), _2_ (2003)
- _baseEncoding_ (string) - ISO8583 message encoding; __Valid values__:
 _ascii_, _binary_
- _fieldFormat_ (object) - ISO8583 field format definition to overwrite the
 default format; **NOTE: for more information please check _Field format
 definition_ above**
- _networkCodeField_ (integer) - ???
- _mtidRouteMap_ (???) - ???

---


CHANGELOG.md
## [7.0.6](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.5...v7.0.6) (2022-06-03)



## [7.0.5](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.4...v7.0.5) (2020-04-27)



## [7.0.4](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.3...v7.0.4) (2020-04-09)


### Bug Fixes

* TAP upgrade ([195bcd8](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/195bcd84a3131ce6d97d69bfafa471cee423ac98))
* TAP upgrade ([b83d2e0](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/b83d2e0b5b972d16910225f110faeb7b15113b18))
* TAP upgrade ([ee5bb44](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/ee5bb4439153f2d2b96d9fa2e00e47fd4c916de5))



## [7.0.3](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.2...v7.0.3) (2020-02-24)


### Bug Fixes

* lint ([a598e99](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/a598e993e30973ab82d1f38b2d1f355262314f69))



## [7.0.2](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.1...v7.0.2) (2020-01-17)



## [7.0.1](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v7.0.0...v7.0.1) (2020-01-15)


### Bug Fixes

* quotes in template string ([#25](https://github.com/softwaregroup-bg/ut-codec-iso8583/issues/25)) ([8b1a311](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/8b1a311))



# [7.0.0](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.2.1...v7.0.0) (2019-10-01)



<a name="6.2.1"></a>
## [6.2.1](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.2.0...v6.2.1) (2019-05-30)



<a name="6.2.0"></a>
# [6.2.0](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.11...v6.2.0) (2018-09-17)


### Bug Fixes

* mask some object properties ([465cebe](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/465cebe))



<a name="6.1.11"></a>
## [6.1.11](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.10...v6.1.11) (2018-02-15)


### Bug Fixes

* dependencies ([6728144](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/6728144))



<a name="6.1.10"></a>
## [6.1.10](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.9...v6.1.10) (2018-01-23)


### Bug Fixes

* handle parse error ([#6](https://github.com/softwaregroup-bg/ut-codec-iso8583/issues/6)) ([d333a1f](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/d333a1f))



<a name="6.1.9"></a>
## [6.1.9](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.8...v6.1.9) (2017-12-14)


### Bug Fixes

* handle variable sizes properly ([5030bdc](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/5030bdc))



<a name="6.1.8"></a>
## [6.1.8](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.7...v6.1.8) (2017-12-14)


### Bug Fixes

* revert ([a584720](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/a584720))



<a name="6.1.7"></a>
## [6.1.7](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.6...v6.1.7) (2017-12-13)


### Bug Fixes

* make emvTags to be coded when buffer ([b324592](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/b324592))



<a name="6.1.6"></a>
## [6.1.6](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.5...v6.1.6) (2017-12-05)


### Bug Fixes

* fill missing error codes ([#4](https://github.com/softwaregroup-bg/ut-codec-iso8583/issues/4)) ([c971c43](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/c971c43))



<a name="6.1.5"></a>
## [6.1.5](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.4...v6.1.5) (2017-11-24)


### Bug Fixes

* emv ([b996555](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/b996555))
* enhance stan usability ([826049d](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/826049d))
* parse emv tags in error response ([0c7d842](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/0c7d842))



<a name="6.1.4"></a>
## [6.1.4](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.3...v6.1.4) (2017-11-20)


### Bug Fixes

* emv tags ([d941081](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/d941081))



<a name="6.1.3"></a>
## [6.1.3](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.2...v6.1.3) (2017-11-18)


### Bug Fixes

* use defineError passed from port ([0881e75](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/0881e75))



<a name="6.1.2"></a>
## [6.1.2](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.1...v6.1.2) (2017-11-13)


### Bug Fixes

* enhance stan usability ([#1](https://github.com/softwaregroup-bg/ut-codec-iso8583/issues/1)) ([1053b49](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/1053b49))



<a name="6.1.1"></a>
## [6.1.1](https://github.com/softwaregroup-bg/ut-codec-iso8583/compare/v6.1.0...v6.1.1) (2017-10-24)


### Bug Fixes

* allow easier updates till we go in production ([41a23c9](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/41a23c9))



<a name="6.1.0"></a>
# 6.1.0 (2017-10-23)


### Bug Fixes

* fix lint error ([6bc9e06](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/6bc9e06))
* **codec:** change hex to binhex in default field format ([814f54f](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/814f54f))
* add dependency ([e162961](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/e162961))
* add error codes ([a46b003](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/a46b003))
* add missing dependency ([3c304ae](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/3c304ae))
* add mtid in smpp.decode ([f685850](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/f685850))
* add sensors to the session, handle mac ([c9c8b40](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/c9c8b40))
* arqc format ([ffd1ce1](https://github.com/softwaregroup-bg/ut-codec-iso8583/commit/ffd1ce1))

test/bitmaps.js
const tap = require('tap');
const {define, get, fetch} = require('ut-unittest/errorApi.js')();
const errorApi = { getError: get, fetchErrors: fetch, defineError: define };
const config = require('./config/testData')();

const ISO8583Codec = require('../index');
const iso8583 = new ISO8583Codec(Object.assign({}, config.messageFormat, errorApi));

const obj = config.bitmaps.testObjects;
const buff = config.bitmaps.testBuffers;

tap.test('Test bitmaps generation and encoding', (t) => {
    t.same(iso8583.encode(obj.bmp1, {}, {}, {}).toString('ascii').slice(4, 20), buff.bmp1, 'Encoding of bitmap 1');
    t.same(iso8583.encode(obj.bmp1and2, {}, {}, {}).toString('ascii').slice(4, 36), buff.bmp1and2, 'Encoding of bitmap 1 and 2');
    t.same(iso8583.encode(obj.bmpFull, {}, {}, {}).toString('ascii').slice(4, 36), buff.bmpFull, 'Encoding of bitmap 1 and 2 with all fields');
    t.same(iso8583.encode(obj.bmpEmpty, {}, {}, {}).toString('ascii').slice(4, 20), buff.bmpEmptyDefault, 'Encoding of bitmap with no fields');
    t.end();
});
tap.test('Test bitmaps decoding', (t) => {
    let tmp = iso8583.decode(Buffer.from(buff.bmp1msg, 'hex'), {}, {}, {});
    t.same(tmp[0].toString('hex'), buff.bmp1, 'Decoding of bitmap 1');
    tmp = iso8583.decode(Buffer.from(buff.bmp1and2msg, 'hex'), {}, {}, {});
    t.same(tmp[0].toString('hex') + tmp[1].toString('hex'), buff.bmp1and2.toLocaleLowerCase(), 'Decoding of bitmap 1 and 2');
    tmp = iso8583.decode(Buffer.from(buff.bmpFullmsg, 'hex'), {}, {}, {});
    t.same(tmp[0].toString('hex') + tmp[1].toString('hex'), buff.bmpFull.toLocaleLowerCase(), 'Decoding of bitmap 1 and 2 with all fields');
    tmp = iso8583.decode(Buffer.from(buff.bmpEmptymsg, 'hex'), {}, {}, {});
    t.same(tmp[0].toString('hex'), buff.bmpEmpty, 'Decoding of bitmap with no fields');
    t.end();
});


test/decoding.test.js
const tap = require('tap');
const {define, get, fetch} = require('ut-unittest/errorApi.js')();
const errorApi = { getError: get, fetchErrors: fetch, defineError: define };
const config = require('./config/testData')();

const ISO8583Codec = require('../index');
const iso8583 = new ISO8583Codec(Object.assign({}, config.messageFormat, errorApi));

const obj = config.decoding.testObjects;
const buff = config.decoding.testBuffers;
tap.test('decode correct buffer to message object', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffCorrect, 'hex'), meta, context, log);
    t.same(objDecode, obj.messageCorrect, 'test message object');
    t.same(meta, obj.metaCorrect, 'test meta object');
    t.same(context, obj.empty, 'test context object');
    t.end();
});
tap.test('decode buffer with wrong frame', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffWrongFrame, 'hex'), meta, context, log);
    t.match(objDecode, obj.messageParserError, 'test message object');
    t.match(meta, obj.metaError, 'test meta object');
    t.end();
});
tap.test('decode buffer with extra bytes', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffExtralong, 'hex'), meta, context, log);
    t.match(objDecode, obj.messageParserError, 'test message object');
    t.match(meta, obj.metaError, 'test meta object');
    t.end();
});
tap.test('decode buffer with wrong field', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffWrongField, 'ascii'), meta, context, log);
    t.match(objDecode, obj.messageParserError, 'test message object');
    t.match(meta, obj.metaError, 'test meta object');
    t.end();
});
tap.test('decode buffer ISO field 39 error', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffISOError, 'hex'), meta, context, log);
    t.match(objDecode, obj.messageGenericError15, 'test message object');
    t.match(meta, obj.metaErrorGeneric, 'test meta object');
    t.end();
});
tap.test('decode buffer with MTI 0800', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    const objDecode = iso8583.decode(Buffer.from(buff.buffMTI0800, 'hex'), meta, context, log);
    t.match(objDecode, obj.messageMTI0800, 'test message object');
    t.match(meta, obj.metaMTI0800, 'test meta object');
    t.end();
});


test/encoding.test.js
const tap = require('tap');
const {define, get, fetch} = require('ut-unittest/errorApi.js')();
const errorApi = { getError: get, fetchErrors: fetch, defineError: define };
const config = require('./config/testData')();

const ISO8583Codec = require('../index');
const iso8583 = new ISO8583Codec(Object.assign({}, config.messageFormat, errorApi));

const obj = config.encoding.testObjects;
const buff = config.encoding.testBuffers;

tap.test('encode message object to buffer - all fields', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    t.same(iso8583.encode(obj.allFields, meta, context, log), Buffer.from(buff.allFields, 'hex'), 'encode all fields');
    t.same(meta, obj.metaAllFields, 'test meta object');
    t.same(context, obj.empty, 'test context object');
    t.end();
});
tap.test('encode message object without field 11 to buffer - test context', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    t.same(iso8583.encode(obj.withoutField11, meta, context, log), Buffer.from(buff.withoutField11, 'hex'), 'encode all fields');
    t.same(meta, obj.metaWithoutField11, 'test meta object');
    t.same(context, obj.contextWithoutField11, 'test meta object');
    t.end();
});
tap.test('encode message object to buffer - all fields + EMV tags', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    t.same(iso8583.encode(obj.allFieldsEMV, meta, context, log), Buffer.from(buff.allFieldsEMV, 'hex'), 'encode all fields');
    t.same(meta, obj.metaAllFieldsEMV, 'test meta object');
    t.same(context, obj.empty, 'test context object');
    t.end();
});
tap.test('encode message object to buffer - without field 11, max trace', (t) => {
    const meta = {};
    const context = {
        trace: 1000000
    };
    const log = {};
    t.same(iso8583.encode(obj.withoutField11MaxTrace, meta, context, log), Buffer.from(buff.withoutField11, 'hex'), 'encode all fields');
    t.same(meta, obj.metaWithoutField11, 'test meta object');
    t.same(context, obj.maxTrace, 'test context object');
    t.end();
});
iso8583.fieldFormat.footer = {size: 8, name: 'Footer', format: 'string'};
iso8583.fieldFormat.header = {size: 8, name: 'Header', format: 'string'};
tap.test('encode message object to buffer - all fields + header +  footer', (t) => {
    const meta = {};
    const context = {};
    const log = {};
    t.same(iso8583.encode(obj.allFields, meta, context, log), Buffer.from('00'.repeat(8) + buff.allFields + '00'.repeat(8), 'hex'), 'encode all fields');
    t.same(meta, obj.metaAllFields, 'test meta object');
    t.same(context, obj.empty, 'test context object');
    t.end();
});


index.js
'use strict';
const merge = require('lodash.merge');
const defaultFields = require('./fields');
const bitSyntax = require('ut-bitsyntax');
const template = require('ut-function.template');
const maskSymbol = Buffer.from('*', 'ascii').toString('hex');

function convertError(msg) {
    return Object.keys(msg).reduce((prev, current) => {
        prev[(/^[^A-Za-z_]/.test(current) ? 'iso' : '') + current] = msg[current];
        return prev;
    }, {iso: msg});
}

function getFormat(format, fallback) {
    return (format && {numeric: 'string-left-zero', string: 'string-right-space', amount: 'string-left-zero', bcdamount: 'string'}[format]) || format || fallback || 'binary';
}

const getMaskList = (arr, objArr) => {
    return arr
        .filter((v) => objArr[v.split(':').shift()])
        .map((v) => {
            let enc = 'ascii';
            if (v.endsWith(':hex')) {
                enc = 'hex';
                v = v.split(':').shift();
            }
            return Buffer.from(objArr[v], enc).toString('hex');
        });
};

const decodeBufferMask = (maskFields) => (buffer, messageParsed) => {
    const maskList = getMaskList(maskFields, messageParsed);

    if (maskList.length) { // TODO: set some emv default field
        const newBuffer = maskList.reduce((a, cur) => {
            return a.split(cur).join((new Array(cur.length / 2)).fill(maskSymbol).join(''));
        }, buffer.toString('hex'));

        return Buffer.from(newBuffer, 'hex');
    }
    return buffer;
};

const encodeBufferMask = (maskFields) => (buffer, message) => {
    const maskList = getMaskList(maskFields, message);
    if (maskList.length) {
        const newBuffer = maskList.reduce((a, cur) => {
            return a.split(cur).join((new Array(cur.length / 2)).fill(maskSymbol).join(''));
        }, buffer.toString('hex'));

        return Buffer.from(newBuffer, 'hex');
    }
    return buffer;
};

function Iso8583(config) {
    if (!config.defineError) {
        throw new Error('Missing config.defineError, check if are you using latest version of ut-port-tcp.');
    }
    if (!config.getError) {
        throw new Error('Missing config.getError, check if are you using latest version of ut-port-tcp.');
    }
    if (!config.fetchErrors) {
        throw new Error('Missing config.fetchErrors, check if are you using latest version of ut-port-tcp.');
    }
    const maskFields = config.maskFields || ['2', '35'];
    this.traceTemplate = template(config.traceTemplate || '${(message.mtid || \'00\').substr(0, 2)}${message[11]}', ['message']); // eslint-disable-line no-template-curly-in-string
    this.errors = require('./errors')(config);
    this.decodeBufferMask = decodeBufferMask(maskFields);
    this.encodeBufferMask = encodeBufferMask(maskFields);
    this.networkCodes = Object.assign({
        '001': 'keyChange',
        '002': 'signOff',
        '061': 'echo',
        161: 'keyChange',
        201: 'cutOver',
        301: 'echo'
    }, config.networkCodes);
    this.emvParser = config.emvParser || require('ut-emv');
    this.successResponseIdentifier = config.successResponseIdentifier || '00';
    this.fieldFormat = merge({}, defaultFields[(config.version || '0') + (config.baseEncoding || 'ascii')], config.fieldFormat);
    this.framePattern = bitSyntax.matcher('header:' + this.fieldFormat.header.size + '/' + getFormat(this.fieldFormat.header.format) +
        ', mtid:' + this.fieldFormat.mtid.size + '/' + getFormat(this.fieldFormat.mtid.format) +
        ', field0:' + this.fieldFormat['0'].size + '/' + getFormat(this.fieldFormat['0'].format) +
        ', rest/binary');
    this.fieldPatterns = [];
    this.fieldBuilders = [bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat['0'].format))];
    this.fieldBuilders.mtid = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.mtid.format));
    this.fieldBuilders.header = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.header.format));
    this.fieldBuilders.footer = bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat.footer.format));
    this.prefixBuilders = [null];
    this.footerMatcher = bitSyntax.matcher('footer:' + this.fieldFormat.footer.size + '/' + getFormat(this.fieldFormat.footer.format) + ', rest/binary');
    this.mtidRouteMap = {...{decode: {}, encode: {}}, ...config.mtidRouteMap};
    this.networkCodeField = config.networkCodeField || 70;
    let group = 0;
    while (this.fieldFormat[(group + 1) * 64]) {
        const pattern = [];
        for (let i = 1; i <= 64; i += 1) {
            const field = group * 64 + i;
            if (this.fieldFormat[field].prefixSize) { // if the field is with variable size
                const factor = this.fieldFormat[field].prefixFactor;
                pattern.push('prefix' + field + ':field' + field + 'Size/' + getFormat(this.fieldFormat[field].prefixFormat, 'string-left-zero') +
                    ', field' + field + ':prefix' + field + '/' + getFormat(this.fieldFormat[field].format) + (factor ? '-unit:' + (8 / factor) : ''));
                this.prefixBuilders.push(bitSyntax.parse('prefix:' + this.fieldFormat[field].prefixSize + '/' +
                    getFormat(this.fieldFormat[field].prefixFormat, 'string-left-zero')));
            } else { // if the field is with fixed size
                pattern.push('field' + field + ':field' + field + 'Size/' + getFormat(this.fieldFormat[field].format));
                this.prefixBuilders.push(null);
            }
            this.fieldBuilders.push(bitSyntax.parse('field:fieldSize/' + getFormat(this.fieldFormat[field].format)));
        }
        pattern.push('rest/binary');
        this.fieldPatterns.push(bitSyntax.matcher(pattern.join(', ')));
        group += 1;
    }
}

Iso8583.prototype.fieldSizes = function(bitmap, start) {
    /* jshint bitwise: false */
    const result = {};
    for (let i = 0; i <= 63; i += 1) {
        const size = bitmap && ((bitmap[i >> 3] & (128 >> (i % 8))) !== 0);
        if (size) {
            result['field' + (start + i) + 'Size'] = this.fieldFormat[start + i].prefixSize || this.fieldFormat[start + i].size;
        } else {
            result['field' + (start + i) + 'Size'] = 0;
        }
    }
    return result;
};

Iso8583.prototype.decode = function(buffer, $meta, context, log) {
    let internalError = false;
    let message = {};
    try {
        if (log && log.trace) {
            const bufferMasked = this.decodeBufferMask(buffer, message);
            log.trace({$meta: {mtid: 'frame', method: 'iso8583.decode'}, message: bufferMasked, log: context && context.session && context.session.log});
        }
        let frame = this.framePattern(buffer);
        let bitmapField = 0;
        if (frame) {
            message = {header: frame.header, mtid: frame.mtid, 0: frame.field0};
            let parsedLength = buffer.length - frame.rest.length;
            let group = 0;
            while (frame) {
                const fieldPattern = this.fieldPatterns[group];
                if (!fieldPattern) {
                    if (frame.rest && frame.rest.length) {
                        if (this.fieldFormat.footer && this.fieldFormat.footer.size) {
                            frame = this.footerMatcher(frame.rest || Buffer.alloc(0));
                            message.footer = frame && frame.footer;
                        }
                        if (frame.rest && frame.rest.length) {
                            throw new Error('Not all data was parsed. Remaining ' + frame.rest.length + ' bytes at offset ' + parsedLength +
                            ' starting with 0x' + frame.rest.toString('hex') + '\r\nmessage:' + JSON.stringify(message));
                        }
                    }
                    break;
                }
                const fieldSizes = this.fieldSizes(frame['field' + bitmapField], group * 64 + 1);
                const rest = frame.rest;
                frame = fieldPattern && fieldPattern(rest, fieldSizes);
                if (!frame && fieldPattern) {
                    for (let failField = (group + 1) * 64; failField >= group * 64 + 1; failField -= 1) { // find at which field we failed by skipping fields from the end
                        fieldSizes['field' + failField + 'Size'] = 0;
                        frame = fieldPattern && fieldPattern(rest, fieldSizes);
                        if (frame && frame.rest && frame.rest.length && this.fieldFormat.footer && this.fieldFormat.footer.size) {
                            frame = this.footerMatcher(frame.rest || Buffer.alloc(0));
                            message.footer = frame && frame.footer;
                        }
                        if (frame) {
                            parsedLength += rest.length - frame.rest.length;
                            throw new Error('Parsing failed at field ' + failField + '. Remaining ' + frame.rest.length + ' bytes at offset ' + parsedLength +
                                ' starting with 0x' + frame.rest.toString('hex') + '\r\nmessage:' + JSON.stringify(message));
                        }
                    }
                    throw new Error('Parsing failed at unknown field');
                }
                parsedLength += rest.length - frame.rest.length;
                bitmapField = group * 64 + 1;
                for (let fieldNo = group * 64 + 1; fieldNo <= (group + 1) * 64; fieldNo += 1) {
                    if (fieldSizes['field' + fieldNo + 'Size']) {
                        message[fieldNo] = frame['field' + fieldNo];
                    }
                }
                group += 1;
            }
            message.mtid = this.mtidRouteMap.decode[message.mtid] || message.mtid;
            if (['0800', '0810'].includes(message.mtid)) {
                $meta.opcode = String(message[this.networkCodeField] || '');
                $meta.opcode = this.networkCodes[$meta.opcode] || $meta.opcode;
            } else {
                $meta.opcode = String(message[3] || '').substr(0, 2);
            }
            $meta.trace = this.traceTemplate(message);
            if (message.mtid && message.mtid.slice) {
                $meta.mtid = {
                    0: 'request',
                    1: (message[39] || this.successResponseIdentifier) === this.successResponseIdentifier
                        ? 'response'
                        : 'error',
                    2: 'request',
                    3: (message[39] || this.successResponseIdentifier) === this.successResponseIdentifier
                        ? 'response'
                        : 'error',
                    4: 'notification',
                    5: 'notification'
                }[(message.mtid.slice(-2).substr(0, 1))] || 'error';
            }
            $meta.method = message.mtid + ($meta.opcode ? '.' + $meta.opcode : '');
            if (message[55]) {
                try {
                    message = Object.assign({}, message, {emvTags: this.emvParser.tagsDecode(message[55], {})});
                } catch (e) {
                    $meta.mtid = 'error';
                    internalError = this.errors['iso8583.parser'];
                    message.errorStack = e;
                }
            }
            if ($meta.mtid === 'error') {
                const err = internalError || (this.errors[`iso8583.${message[39]}`] || this.errors['iso8583.generic']);
                message = err(convertError(message));
            }
            // for mac verify
            message.rawData = buffer;

            return message;
        } else {
            throw new Error('Unable to parse message type or first bitmap!');
        }
    } catch (e) {
        $meta.mtid = 'error';
        message.errorStack = e;
        message = this.errors['iso8583.parser'](convertError(message));
        return message;
    }
};

Iso8583.prototype.encodeField = function(fieldName, fieldValue) {
    const prefixBuilder = this.prefixBuilders[fieldName];
    const builder = this.fieldBuilders[fieldName];
    let fieldSize;
    if (!prefixBuilder) {
        fieldSize = this.fieldFormat[fieldName].size;
    } else if (fieldValue == null || !fieldValue.toString) {
        fieldSize = 0;
    } else if (builder && builder[0] && builder[0].binhex) {
        fieldSize = Buffer.byteLength('0' + fieldValue, 'hex');
    } else if (builder && builder[0] && builder[0].hexbin) {
        fieldSize = Buffer.byteLength(fieldValue, 'utf-8') * 2;
    } else {
        fieldSize = fieldValue.toString().length;
    }
    const field = bitSyntax.build(builder, {
        field: fieldValue,
        fieldSize
    });
    const factor = this.fieldFormat[fieldName].prefixFactor || 1;
    let prefix = field.length * factor;
    if (builder[0].binhex && fieldValue.length !== prefix && factor === 2) {
        prefix = fieldValue.length;
    }
    return prefixBuilder ? Buffer.concat([bitSyntax.build(prefixBuilder, {prefix}), field]) : field;
};

Iso8583.prototype.encode = function(message, $meta, context, log) {
    /* jshint bitwise: false */
    const buffers = new Array(64 * this.fieldPatterns.length);
    const emptyBuffer = Buffer.alloc(0);
    if (!message[11]) {
        context.trace = context.trace || 0;
        message[11] = `${'0'.repeat(this.fieldFormat[11].size)}${context.trace}`.slice(-this.fieldFormat[11].size);
        context.trace++;
        if (context.trace >= Math.pow(10, this.fieldFormat[11].size)) {
            context.trace = 0;
        }
    }
    $meta.trace = this.traceTemplate(message);
    if (message.emvTags) {
        message[55] = this.emvParser.tagsEncode(message.emvTags);
    }
    const bitmaps = Array.apply(null, new Array(8 * this.fieldPatterns.length)).map(Number.prototype.valueOf, 0); // zero filled array
    for (let i = 64 * this.fieldPatterns.length; i >= 0; i -= 1) {
        if (i === 0) {
            buffers[i] = this.encodeField(i, Buffer.from(bitmaps.slice(0, 8)));
        } else if (i % 64 === 1 && i < 64 * (this.fieldPatterns.length - 1)) {
            const index = (i >> 6) << 3;
            const bitmap = bitmaps.slice(index + 8, index + 16);
            if (bitmap.reduce(function(p, n) { return p + n; })) {
                bitmaps[(i - 1) >> 3] |= (128 >> (i - 1) % 8);
                buffers[i] = this.encodeField(i, Buffer.from(bitmap));
            } else {
                buffers[i] = emptyBuffer;
            }
        } else if (message[i] != null) {
            bitmaps[(i - 1) >> 3] |= (128 >> (i - 1) % 8);
            buffers[i] = this.encodeField(i, message[i]);
        } else {
            buffers[i] = emptyBuffer;
        }
    }

    message.mtid = this.mtidRouteMap.encode[message.mtid] || message.mtid;
    buffers.unshift(this.encodeField('mtid', message.mtid || Buffer.alloc(0)));

    if (this.fieldFormat.header && this.fieldFormat.header.size) {
        buffers.unshift(this.encodeField('header', message.header || Buffer.alloc(this.fieldFormat.header.size)));
    }
    if (this.fieldFormat.footer && this.fieldFormat.footer.size) {
        buffers.push(this.encodeField('footer', message.footer || Buffer.alloc(this.fieldFormat.footer.size)));
    }
    const buffer = Buffer.concat(buffers);
    if (log && log.trace) {
        const bufferMasked = this.encodeBufferMask(buffer, message);
        log.trace({$meta: {mtid: 'frame', method: 'iso8583.encode'}, message: bufferMasked, log: context && context.session && context.session.log});
    }
    return buffer;
};

module.exports = Iso8583;
